rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+use cases listed above.
+
+#### Scaffolding Templates {#scaffolding-templates}
+
+The scaffolding template CLI command itself is still **experimental**, therefore
+excluded from compatibility compliance. Note that generated pipeline code &
+config are considered user code, therefore they can expect the same
+compatibility compliance as listed above.
+
+Putting it differently:
+
+*   The structure of code and configs generated is subject to (possibly
+    breaking) change between minor versions;
+*   Once generated, the produced pipeline code and configs are essentially user
+    code which relies on TFX’s public API, thus guaranteed to work with newer
+    minor versions within the same major version, as explained in backward
+    compatibility rules.
+    *   This rule ensures that even if a user upgraded the `tfx` package in
+        their environment, the scaffolded pipeline can still function.
+
+#### Examples, Notebooks and Tutorials {#examples-notebooks-and-tutorials}
+
+Examples, Notebooks (Colab) and Tutorials are **excluded** from the
+compatibility compliance. These are created mostly for demonstrating how to use
+TFX in accordance with best practices, therefore often subject to a particular
+version of TFX.
+
+These examples should use the stable public API of `tfx` as much as possible.
+Usage of experimental or deprecated APIs in these examples should be explicitly
+called out, and the examples should pin their dependency version of `tfx` down
+to a narrow range as much as possible in these rare cases.
+
+#### New Orchestrator {#new-orchestrator}
+
+The new orchestrator is in early stages of development and we don’t think it’ll
+be production-ready for 1.0 timeline, thus it will enter 1.0 as `experimental`.
+
+#### Portability Layer to Other Runners (aka Orchestrators) {#portability-layer-to-other-runner-orchestrators}
+
+This portability layer in tfx refers to various DagRunners which runs TFX
+pipelines on various external orchestrators, the _portable_ module and the
+abstraction between CLI and Dag runner. These libraries will comply with
+backward compatibility.
+
+Notice that several Runners are experimental by their nature, or slated to be
+deprecated. Specifically, we plan to support the following runners with backward
+compatibility:
+
+*   LocalDagRunner
+*   KubeflowV2DagRunner
+
+The following Dag runners will remain experimental:
+
+*   AirflowDagRunner
+    *   Honestly speaking, support of this one is on the fence. Due to lack of
+        proper configuration for a distributed backend support, it’s not much
+        more powerful than the LocalDagRunner when used out-of-box. However,
+        some partner teams managed to connect this to a distributed backend.
+        Until that is supported in TFX, we don’t think this one is stable.
+
+The following dag runners are slated to be deprecated and therefore won’t
+comply:
+
+*   BeamDagRunner: this will be replaced by the LocaDagRunner
```

Comment:
is the idea to make LocalDagRunner parallelizable for 1.X? (given that Beam is parallizable locally while currently LocalDagRunner it is not).

Reply:
We don't have such a plan right now, although making such a change should be a net improvement w/o breaking change.

Do you have a concrete use case of a parallelized LocalDagRunner?

@charlesccychen

Reply:
Given most of our pipelines just trigger remote jobs, allowing us to trigger several dataflow jobs in parallel for some pre-processing would be nice.

Reply:
I think its fine to leave as is. Would be nice to have a public facing ticket to track the work to make it parallelizable.

For prioritization, LocalDagRunner may not be considered a feasible option unless its parallelizable.